### PR TITLE
feat: Linked Sessions & Conversation Triggers

### PR DESCRIPTION
**Pull request by Pizzaface**
## Linked Sessions & Conversation Triggers

Adds a parent-child session linking system with structured conversation triggers, enabling coordinated multi-agent workflows through the PizzaPi relay. Spawned child sessions are automatically linked to their parent — child events (questions, plans, completion) surface as trigger messages in the parent's conversation for approval, escalation, or follow-up.

### What's New

**Protocol** (`packages/protocol`)
- `parentSessionId` field on session data (`shared.ts`, `runner.ts`)
- `session_trigger` and `trigger_response` relay events with typed payloads (`relay.ts`)
- `trigger_response` viewer event for web-based trigger handling (`viewer.ts`)

**Server** (`packages/server`)
- Redis-backed child session index with `addChildSession` / `getChildSessions` / `getDescendantSessionIds` helpers (with cycle-detection via visited set)
- `session_trigger` + `trigger_response` routing in the relay namespace with **user ownership validation** on both directions
- `trigger_response` handler in viewer namespace for web UI responses
- `parentSessionId` included in `session_added` hub broadcasts and session registration
- Cross-node delivery support for triggers via Redis pub/sub

**CLI — Triggers Extension** (`packages/cli/src/extensions/triggers/`)
- New extension with `tell_child`, `respond_to_trigger`, and `escalate_trigger` tools
- `tell_child` delivers as agent input (not passive message bus) so the child actually acts on it
- Conversation trigger type registry with renderers for `plan_review`, `question`, `approval`, `error`, `status_update`, and `session_complete`
- `AskUserQuestion` intercept for child sessions — routes questions to parent instead of blocking
- `plan_mode` intercept for child sessions — routes plan review to parent with `approve` / `cancel` / `edit` actions
- `session_complete` trigger fired on child exit (on `agent_end`, not just `session_shutdown`) with last assistant message included in summary
- Session complete summary capped at 16K chars; overflow saved to file and path included in trigger
- `session_complete` triggers are respondable — parent can `ack` or `followUp` with more work
- 10-minute follow-up grace period for child sessions after `session_complete`
- Fire-and-forget triggers (e.g., shutdown) skip tracking in `receivedTriggers`
- `spawn_session` automatically sets `parentSessionId` to link parent-child

**CLI — Resilience**
- Malformed agent/skill/plugin `.md` files no longer crash session startup — bad files are skipped with a warning
- Covers broken symlinks, binary content, and malformed YAML frontmatter

**CLI — Other**
- "End Session Group" option when ending parent sessions — terminates all descendants
- Built-in system prompt updated with trigger system guidance (don't stall waiting for child triggers, use `respond_to_trigger` / `escalate_trigger`)

**UI** (`packages/ui`)
- **Session tree**: parent-child grouping in sidebar with indented child sessions, auto-expand on new child, real-time updates on `session_added` with `parentSessionId`
- **Trigger cards**: standalone `TriggerCard` component rendering plan reviews, questions, approvals, errors, status updates, and session complete summaries — with inline response actions
- **Tool renderers**: custom cards for `tell_child`, `respond_to_trigger`, `escalate_trigger`, `spawn_session` tools showing structured trigger/response data
- **Inter-agent message cards**: enhanced rendering for message bus events

**Docs** (`packages/docs`, `AGENTS.md`)
- Runner daemon docs updated with `parentSessionId` spawn parameter
- Environment variables reference updated with trigger-related config
- `AGENTS.md` updated with trigger-based parent-child interaction model
- Design spec: `specs/designs/2026-03-13-linked-sessions-conversation-triggers-design.md`
- Implementation plan: `specs/plans/2026-03-13-linked-sessions-triggers.md`

### Testing

| Area | Tests |
|------|-------|
| Trigger registry & renderers | Unit tests for all trigger types, response parsing, renderer selection |
| Trigger routing | Integration tests for full relay round-trip (parent → child → response) |
| Child session Redis helpers | Unit tests for add/get/getDescendants with cycle detection |
| Session tree (UI) | Unit tests for tree building, sorting, parent lookup |
| TriggerCard parsing (UI) | Unit tests for all trigger type rendering and edge cases |
| Resilient file loading | Tests for broken symlinks, binary content, malformed frontmatter in agents/skills/plugins |

### Security

- User ownership validated on `session_trigger` (sender must own source session)
- User ownership validated on `trigger_response` (responder must own target session)  
- `parentSessionId` ownership validated before linking during session registration
- `parentSessionId` validated before forwarding to runner on spawn

### CODE REVIEW

File: packages/server/src/routes/runners.ts [hunk 1]
```diff
@@ -112,6 +115,7 @@ export const handleRunnersRoute: RouteHandler = async (req, url) => {
                 ...(requestedModel ? { model: requestedModel } : {}),
                 ...(hiddenModels.length > 0 ? { hiddenModels } : {}),
                 ...(requestedAgent ? { agent: requestedAgent } : {}),
+                ...(requestedParentSessionId ? { parentSessionId: requestedParentSessionId } : {}),
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/routes/runners.ts]: Fixed in e3bdabf. Moved the parent session ownership validation **before** the `runnerSocket.emit("new_session", ...)` call. The worker now only receives `parentSessionId` if it passes the same-user check, so invalid/cross-user parent IDs never activate child trigger mode.

File: packages/cli/src/extensions/factories.ts [hunk 2]
```diff
@@ -49,6 +50,7 @@ export function buildPizzaPiExtensionFactories(options: BuildExtensionFactoriesO
         sessionMessagingExtension,
         subagentExtension,
         planModeToggleExtension,
+        triggersExtension,
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/cli/src/extensions/factories.ts]: Fixed in e3bdabf. Moved `triggersExtension` before `remoteExtension` in the factory registration order. Now the `session_shutdown` handler in triggers fires first (sending the `session_complete` trigger via the still-connected relay), then remote's handler disconnects the socket.

File: packages/cli/src/extensions/triggers/extension.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                return { content: [{ type: "text" as const, text: `Error: No pending trigger with ID ${params.triggerId}.` }], details: null as any };
+            }
+            // Fire an escalate trigger to the parent's own session so the web UI
+            // surfaces it to the human viewer (not back to the child).
+            conn.socket.emit("session_trigger" as any, {
+                token: conn.token,
+                trigger: {
+                    type: "escalate",
+                    sourceSessionId: pending.sourceSessionId,
+                    targetSessionId: ownSessionId ?? "",
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/cli/src/extensions/triggers/extension.ts]: Fixed in 01926cb. Replaced `process.env.PIZZAPI_SESSION_ID` with `getRelaySessionId()` (new export from `remote.ts`) which returns the server-assigned relay session ID — works for both runner-spawned and standalone CLI sessions.

File: packages/cli/src/extensions/triggers/extension.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            conn.socket.emit("session_trigger" as any, {
+                token: conn.token,
+                trigger: {
+                    type: "session_complete",
+                    sourceSessionId: ownSessionId ?? "",
+                    sourceSessionName: undefined,
+                    targetSessionId: parentSessionId,
+                    payload: { summary: "Session completed", exitCode: 0 },
+                    deliverAs: "followUp" as const,
+                    expectsResponse: true,
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/cli/src/extensions/triggers/extension.ts]: Fixed in 01926cb. Changed `expectsResponse` to `false` since the child is shutting down and cannot wait for a response. The trigger is now fire-and-forget.

File: packages/cli/src/extensions/remote.ts [hunk 5]
```diff
@@ -1919,6 +1944,20 @@ export const remoteExtension: ExtensionFactory = (pi) => {
             });
         });
 
+        // ── session_trigger — receive triggers from child sessions ─────────
+        sock.on("session_trigger" as any, (data: { trigger: import("./triggers/types.js").ConversationTrigger }) => {
+            const trigger = data?.trigger;
+            if (!trigger) return;
+
+            // Track the trigger for response routing (used by respond_to_trigger tool)
+            trackReceivedTrigger(trigger.triggerId, trigger.sourceSessionId, trigger.type);
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/cli/src/extensions/remote.ts:1982]: Fixed in 4669975. The `session_trigger` handler now checks `trigger.expectsResponse !== false` before calling `trackReceivedTrigger`. Fire-and-forget triggers like `session_complete` are no longer added to `receivedTriggers`, so `respond_to_trigger` correctly reports "no pending trigger" instead of silently dropping the response.

**Comment by Pizzaface** [comment]
This is by design — fire-and-forget triggers (`expectsResponse: false`) like `session_complete` aren't tracked in `receivedTriggers` at all, so stale entries from those can't accumulate. For response-expecting triggers, the child has its own timeout and the worst case of a late viewer click is a no-op on the child side (it already resolved its promise). Adding TTL validation would add complexity without meaningful safety improvement.

**Comment by Pizzaface** [comment]
Fixed in e6be892. Added a `visited` set to `getDescendantSessionIds` BFS traversal — the root `sessionId` is pre-seeded, and each child is checked against the set before enqueueing. This prevents infinite loops from cyclic `parentSessionId` data (including self-references). Added 2 test cases covering both scenarios.

File: packages/cli/src/extensions/remote.ts [hunk 6]
```diff
@@ -145,6 +163,10 @@ export const remoteExtension: ExtensionFactory = (pi) => {
         ? process.env.PIZZAPI_SESSION_ID.trim()
         : randomUUID();
 
+    // ── Child session detection for trigger system ────────────────────────
+    const parentSessionId = process.env.PIZZAPI_WORKER_PARENT_SESSION_ID ?? null;
+    const isChildSession = parentSessionId !== null;
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/cli/src/extensions/remote.ts]: Fixed. Made `parentSessionId` and `isChildSession` mutable (`let`). The `registered` handler now reconciles with the server's confirmed `data.parentSessionId` — if the server rejects the parent link, both variables are cleared so AskUserQuestion/plan_mode fall back to normal local interaction instead of routing through triggers that will time out.

File: packages/cli/src/extensions/remote.ts [hunk 7]
```diff
@@ -2139,6 +2283,64 @@ export const remoteExtension: ExtensionFactory = (pi) => {
...
+                    triggerId,
+                    timeoutMs: 300_000,
+                    ts: new Date().toISOString(),
+                };
+
+                sioSocket.emit("session_trigger", { token: relay.token, trigger });
+
+                // Wait for trigger_response with matching triggerId
+                const triggerResult = await new Promise<{ response: string; cancelled: boolean }>((resolve) => {
+                    const timeout = setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/cli/src/extensions/remote.ts:2328]: Fixed. Both ask_user_question and plan_mode trigger wait blocks now listen for `session_message_error` alongside `trigger_response`. When the relay reports delivery failure (e.g. parent disconnected), the child resolves immediately instead of hanging for 5 minutes.

File: packages/server/src/ws/namespaces/relay.ts [hunk 8]
```diff
@@ -387,6 +403,130 @@ export function registerRelayNamespace(io: SocketIOServer): void {
...
+            // Enforce parent-child linkage: the sender must be the target's parent,
+            // or the target must be a child of the sender. This prevents any relay
+            // client under the same account from injecting trigger responses into
+            // unrelated child sessions.
+            const isParentOfTarget = targetSession.parentSessionId === socket.data.sessionId;
+            const isChildOfSender = senderSession.parentSessionId === targetSessionId;
+            if (!isParentOfTarget && !isChildOfSender) {
+                // Also check the Redis children set as a fallback
+                const children = await getChildSessions(socket.data.sessionId);
+                if (!children.includes(targetSessionId)) {
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/ws/namespaces/relay.ts]: Fixed. Removed the `getChildSessions` fallback from trigger_response auth — now relies solely on live `parentSessionId` linkage. Also added `removeChildSession` cleanup on disconnect so the Redis children set stays current.

File: packages/ui/src/lib/session-tree.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+      const bTime = new Date(b.session.startedAt).getTime();
+      return aTime - bTime;
+    });
+  }
+
+  // Return root sessions: those with no parent, or whose parent is not in this subset
+  const roots: SessionTreeNode[] = [];
+  for (const session of sessions) {
+    if (!session.parentSessionId || !nodeMap.has(session.parentSessionId)) {
+      const node = nodeMap.get(session.sessionId);
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/ui/src/lib/session-tree.ts]: Fixed. Added cycle detection in `buildSessionTree` — cycle members have their parent edges broken and are treated as roots, preventing sessions from disappearing from the sidebar. Added tests for A↔B cycles and self-references.

File: packages/server/src/ws/namespaces/relay.ts [hunk 10]
```diff
@@ -374,11 +378,23 @@ export function registerRelayNamespace(io: SocketIOServer): void {
...
-                targetSocket.emit("session_message" as string, {
-                    fromSessionId: sessionId,
-                    message: messageText,
-                    ts: new Date().toISOString(),
-                });
+                if (data.deliverAs === "input") {
+                    // Deliver as agent input — starts a new turn (used by tell_child).
+                    // Mirrors the viewer namespace "input" handler behavior.
+                    targetSocket.emit("input" as string, {
+                        text: messageText,
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/ws/namespaces/relay.ts:396]: Fixed. Added same-user ownership validation (`senderSession.userId === targetSession.userId`) to the `session_message` handler before forwarding, especially critical for `deliverAs:"input"` which starts new agent turns.

File: packages/ui/src/components/SessionSidebar.tsx [hunk 11]
```diff
@@ -1021,8 +1068,58 @@ export const SessionSidebar = React.memo(function SessionSidebar({
...
                                                     >
+                                                        {/* Expand/collapse toggle for parent sessions */}
+                                                        {(() => {
+                                                            const childCount = childrenByParent.get(s.sessionId) ?? 0;
+                                                            if (childCount > 0) {
+                                                              return (
+                                                                <span
+                                                                  role="button"
+                                                                  tabIndex={-1}
+                                                                  onClick={(e) => {
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/ui/src/components/SessionSidebar.tsx:1082]: Fixed. Changed `tabIndex={-1}` to `tabIndex={0}` on the tree expand/collapse toggle so keyboard users can focus and activate it.
